File: app/src/main/java/com/thewizrd/simpleweather/preferences/CustomPreferenceFragmentCompat.java
package com.thewizrd.simpleweather.preferences;

import android.content.Context;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.thewizrd.shared_resources.lifecycle.LifecycleAwarePreferenceFragmentCompat;
import com.thewizrd.shared_resources.utils.SettingsManager;
import com.thewizrd.simpleweather.App;
import com.thewizrd.simpleweather.snackbar.SnackbarManager;
import com.thewizrd.simpleweather.snackbar.SnackbarManagerInterface;

public abstract class CustomPreferenceFragmentCompat extends LifecycleAwarePreferenceFragmentCompat implements SnackbarManagerInterface {

    private AppCompatActivity mActivity;
    private SnackbarManager mSnackMgr;
    private final SettingsManager settingsMgr = App.getInstance().getSettingsManager();

    public final AppCompatActivity getAppCompatActivity() {
        return mActivity;
    }

    protected final SettingsManager getSettingsManager() {
        return settingsMgr;
    }

    @Nullable
    public abstract SnackbarManager createSnackManager();

    @CallSuper
    public final void initSnackManager() {
        if (mSnackMgr == null) {
            mSnackMgr = createSnackManager();
        }
    }

    public void showSnackbar(@NonNull final com.thewizrd.simpleweather.snackbar.Snackbar snackbar) {
        showSnackbar(snackbar, null);
    }

    @Override
    public void showSnackbar(@NonNull final com.thewizrd.simpleweather.snackbar.Snackbar snackbar, final com.google.android.material.snackbar.Snackbar.Callback callback) {
        runWithView(() -> {
            if (mSnackMgr == null && isAlive()) {
                mSnackMgr = createSnackManager();
            }
            if (mSnackMgr != null && mActivity != null) {
                mSnackMgr.show(snackbar, callback);
            }
        });
    }

    @Override
    public void dismissAllSnackbars() {
        runOnUiThread(() -> {
            if (mSnackMgr != null) mSnackMgr.dismissAll();
        });
    }

    @Override
    public void unloadSnackManager() {
        dismissAllSnackbars();
        mSnackMgr = null;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mActivity = (AppCompatActivity) context;
    }

    @Override
    public void onDetach() {
        mActivity = null;
        super.onDetach();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (!isHidden()) {
            initSnackManager();
        } else {
            dismissAllSnackbars();
        }
    }

    @Override
    public void onPause() {
        unloadSnackManager();
        super.onPause();
    }

    @Override
    public void onDestroy() {
        mActivity = null;
        super.onDestroy();
    }
}

File: app/src/main/java/com/thewizrd/simpleweather/preferences/CustomPreferenceFragmentCompat.kt
package com.thewizrd.simpleweather.preferences

import android.app.Activity
import android.widget.Toast
import androidx.annotation.CallSuper
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.lifecycle.lifecycleScope
import com.thewizrd.shared_resources.lifecycle.LifecycleAwarePreferenceFragmentCompat
import com.thewizrd.simpleweather.snackbar.Snackbar
import com.thewizrd.simpleweather.snackbar.SnackbarManager
import com.thewizrd.simpleweather.snackbar.SnackbarManagerInterface
import kotlinx.coroutines.launch
import com.google.android.material.snackbar.Snackbar as MaterialSnackbar

abstract class CustomPreferenceFragmentCompat : LifecycleAwarePreferenceFragmentCompat(),
    SnackbarManagerInterface {
    private var mSnackMgr: SnackbarManager? = null

    abstract override fun createSnackManager(activity: Activity): SnackbarManager?

    @CallSuper
    override fun initSnackManager(activity: Activity) {
        if (mSnackMgr == null) {
            mSnackMgr = createSnackManager(activity)
        }
    }

    fun showSnackbar(snackbar: Snackbar) {
        showSnackbar(snackbar, null)
    }

    override fun showSnackbar(
        snackbar: Snackbar,
        callback: MaterialSnackbar.Callback?
    ) {
        runWithView {
            activity?.let {
                if (isAlive) {
                    if (mSnackMgr == null) {
                        mSnackMgr = createSnackManager(it)
                    }
                    mSnackMgr?.show(snackbar, callback)
                }
            }
        }
    }

    override fun dismissAllSnackbars() {
        runOnUiThread { mSnackMgr?.dismissAll() }
    }

    override fun unloadSnackManager() {
        dismissAllSnackbars()
        mSnackMgr = null
    }

    @IntDef(Toast.LENGTH_SHORT, Toast.LENGTH_LONG)
    @Retention(AnnotationRetention.SOURCE)
    annotation class ToastDuration

    fun showToast(@StringRes resId: Int, @ToastDuration duration: Int) {
        lifecycleScope.launch {
            context?.let {
                Toast.makeText(it, resId, duration).show()
            }
        }
    }

    fun showToast(message: CharSequence?, @ToastDuration duration: Int) {
        lifecycleScope.launch {
            context?.let {
                Toast.makeText(it, message, duration).show()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (!isHidden) {
            initSnackManager(requireActivity())
        } else {
            dismissAllSnackbars()
        }
    }

    override fun onPause() {
        unloadSnackManager()
        super.onPause()
    }
}